### PR TITLE
Partial: Avenge

### PR DESCRIPTION
**Pull request by JacobWoodson**
## Summary
Fixes a parse-fidelity defect on **Avenge**.

**Issue:** Cost-reduction condition "if a player attacked you during their last turn" is dropped (ModifyCost condition=null), so the {2} reduction applies unconditionally instead of only when the attack condition holds.

## Files changed
- C:\Users\jacob\source\repos\phase-card-runs\crates\engine\src\types\ability.rs
- C:\Users\jacob\source\repos\phase-card-runs\crates\engine\src\types\game_state.rs
- C:\Users\jacob\source\repos\phase-card-runs\crates\engine\src\game\turns.rs
- C:\Users\jacob\source\repos\phase-card-runs\crates\engine\src\game\layers.rs
- C:\Users\jacob\source\repos\phase-card-runs\crates\engine\src\game\quantity.rs
- C:\Users\jacob\source\repos\phase-card-runs\crates\engine\src\game\ability_rw.rs
- C:\Users\jacob\source\repos\phase-card-runs\crates\engine\src\game\ability_scan.rs
- C:\Users\jacob\source\repos\phase-card-runs\crates\engine\src\game\coverage.rs
- C:\Users\jacob\source\repos\phase-card-runs\crates\engine\src\parser\oracle_nom\condition.rs
- C:\Users\jacob\source\repos\phase-card-runs\crates\engine\src\parser\oracle_trigger.rs
- C:\Users\jacob\source\repos\phase-card-runs\crates\engine\src\parser\oracle_effect\conditions.rs
- C:\Users\jacob\source\repos\phase-card-runs\crates\engine\src\parser\oracle_condition.rs
- C:\Users\jacob\source\repos\phase-card-runs\crates\engine\src\parser\oracle_static\tests.rs
- C:\Users\jacob\source\repos\phase-card-runs\crates\engine\src\game\casting_tests.rs

## CR references
- CR 508.6
- CR 514.2
- CR 514.3
- CR 109.5

## Track
Developer

## LLM
Model: claude-opus-4-8
Thinking: high

Tier: Frontier

## Verification
  - `cargo fmt --all` — pass (clean)
  - `./scripts/check-parser-combinators.sh` — pass (Gate G PASS + Gate A PASS; Family D NOT skipped — ran via working msys64 python3 3.9.7, detector self-test suite 10/10 ok; WindowsApps python3 stub bypassed by removing it from PATH)
  - `cargo clippy-strict` — pass (exit 0, no warnings)
  - `cargo test -p phase-engine` — fail (18475 passed, 1 failed: stage2_injector_tests::the_cr_603_5_prompt_census... — Windows-only path-separator artifact, unrelated to Avenge; also cleared a 97GB target/debug/incremental dir that had filled the disk before this run)
  - `cargo export-cards data --stats --sidecar-dir client/public && cp client/public/card-data.json data/card-data.json` — pass (card-data regenerated fresh against this branch; sidecar copied to data/, 98176031 bytes)
  - `cargo coverage` — fail-result (command exit 0, but Avenge supported:false gap_count:1 — Swallow:Condition_If on the cost-reduction clause)
  - `cargo semantic-audit` — pass (command exit 0; 32700 supported audited, 295 flagged; Avenge has 0 findings — it is not in the supported-audit set)

## Scope Expansion
Scope grew by 3 files beyond the plan's list: the plan missed three exhaustive StaticCondition mappers (oracle_trigger.rs, oracle_effect/conditions.rs, oracle_condition.rs) that require a compile-mandatory "-> None" arm.

## Validation Failures
See review/cross-check notes.

## CI Failures
- Avenge is NOT supported: cargo coverage reports supported=false, gap_count=1. Gap = Swallow:Condition_If on 'This spell costs {2} less to cast if a player attacked you during their last turn.' The DestroyAll + GainLife(filtered tracked set: creatures destroyed) body parses correctly, but the conditional cost reduction is swallowed: the ModifyCost static ability captures the {2} reduction with condition=null — the 'if a player attacked you during their last turn' condition is dropped. This is a genuine parse gap surfaced by swallow_check.rs (a file this branch modifies), confirmed against freshly regenerated card-data, not stale data. To reach gap:0 the swallowed StaticCondition must be parsed and attached to the ModifyCost static ability (a feature implementation, out of scope for command-level retry).
- cargo test -p phase-engine: 1 unrelated test failed — game::engine::stage2_injector_tests::the_cr_603_5_prompt_census_is_pinned_so_a_sixth_producer_is_a_counted_event (engine.rs:15185). Windows-only path-separator artifact: file!()/rel.display() yields backslash paths (game\effects\mod.rs:5999) while the test hardcodes forward-slash paths (game/effects/mod.rs:5999). Census content is identical (same files, same lines 5999/6076/9051/452/11427), the count assertions (5 producers, 7 readers, 25 test-lines) all passed, and none of this branch's modified files (triggers.rs, oracle_trigger.rs, swallow_check.rs, integration tests) touch the census sites. Deterministic (retry futile, not a flake); the test's own comment notes it passes on Linux CI. Not caused by this branch and unrelated to Avenge; left unedited (sensitive unrelated census pin, multi-agent-safety).

## Summary by CodeRabbit

* **New Features**
  * Added support for “a player” or “an opponent attacked you during their last turn” conditions.
  * Avenge-style cost reductions now apply when an opponent attacked you during their last turn.
  * Multiplayer games accurately track attacks across opponents and turns.
  * Added readable condition text for supported card rules.

* **Bug Fixes**
  * Corrected attack-history updates, including clearing records after no-attack turns.

* **Tests**
  * Added coverage for parsing, cost reductions, multiplayer behavior, and attack-history persistence.

### CODE REVIEW

**Review by matthewevans — CHANGES_REQUESTED** [review]
**Request changes — the last-turn defender ledger is preserved but not given the CR-required departed-player expiry.**

## 🔴 Blocker

**[HIGH]** Eliminated players are excluded too early from the layer predicate. Evidence: `crates/engine/src/game/layers.rs:1617-1621` filters eliminated players, while `docs/MagicCompRules.txt:6429-6433` (CR 800.4i) says actions during a player's last turn remain findable until that player's next turn would have begun. `types/game_state.rs:14142-14152` and the cleanup in `turns.rs:2148-2161` preserve the record, but introduce no departed-player expiry. Why it matters: the engine loses a required last-turn attack relationship immediately on elimination instead of retaining it through the specified skipped-turn boundary. Suggested fix: carry the relationship through that CR boundary, then expire it when the eliminated player's next turn would begin; add a production-pipeline test for both sides of that boundary.

**[HIGH]** Required CI is failing the CR733 authority census. Evidence: run `30930044012`, job `92062093858`, reports `attacked_defenders_last_turn` as missing. Why it matters: the lifecycle field lacks the required authoritative classification/fixture evidence. Suggested fix: classify it at the authority that owns its lifecycle and update the corresponding fixture consistently with the retention/expiry behavior above.

## ✅ Evidence checked

The current head-bound parse-diff is scoped to Avenge only: one changed static/ReduceCost signature, matching the claimed scope. I also checked CodeRabbit's immediate-elimination suggestion against CR 800.4i; it conflicts with the verified rule and is not the requested fix.

**Recommendation: request changes for the CR 800.4i retention boundary and the CR733 census failure, then re-request review on a new head.**

File: crates/engine/src/game/casting_tests.rs [hunk 1]
```diff
@@ -2766,6 +2766,141 @@ fn visions_of_ruin_flashback_commander_mv_reduces_flashback_cost() {
...
+
+    // Opponents attacked each other but not you ⇒ false (the defender must be you).
+    let mut state = GameState::new(FormatConfig::standard(), 3, 7);
+    state
+        .attacked_defenders_last_turn
+        .insert(PlayerId(1), [PlayerId(2)].into_iter().collect());
+    state
+        .attacked_defenders_last_turn
+        .insert(PlayerId(2), [PlayerId(1)].into_iter().collect());
+    assert!(!evaluate_condition_for_test(&state, &cond, you, src));
```
[the remark this answers is not in the document]
Comment (matthewevans) [reply crates/engine/src/game/casting_tests.rs:3267]: Verified against the Comprehensive Rules: CR 800.4i explicitly allows an effect that requires information about actions players have taken to find actions taken by a player who has left the game. The existing test intentionally keeps P2\x27s recorded attack and asserts the condition remains true after P2 leaves; changing it to false would be rules-incorrect. No code change is appropriate for this thread.

**Comment by matthewevans** [comment]
**Maintainer hold — current head `67d51ffe3290d38b4e008e0c33b380a35f7df160`.**

The source/serializer correction is clean. This branch is behind current `main`; Rust lint and test shards are still in progress, and the available `coverage-parse-diff` artifact is for older head `48ca248…`. Update/rebase as needed, then let exact-current parse evidence and required checks settle; approval/enqueue will resume after that. No author code change is requested at this time.

**Comment by matthewevans** [comment]
**Maintainer hold — current head `6ab4ce9a955fe9e18af56f4d82135087ec43fb92`.**

The current source is clean, but `main` has advanced and this branch is behind it. Rust test shard 1 is still in progress, while the available `coverage-parse-diff` artifact is for older head `67d51…`. Maintainer refresh/update will bring the branch current; then let exact-current parse evidence and required checks settle before approval/enqueue resumes. No author code change is requested at this time.

**Review by matthewevans — APPROVED** [review]
Re-reviewed current main-refreshed head 6ab4ce9: deterministic attack-history serialization and its exact census coverage are complete; CI is green.
